blog/_posts/2017-01-24-cudanative-nvidia.md

```diff
@@ -0,0 +1,234 @@
+with syntax that is familiar to users of other technical computing environments. The
+language has been created with performance in mind, and combines careful language design
+with a sophisticated LLVM-based compiler.
+
+Julia is well regarded for programming multicore CPUs and large parallel computing systems,
+but the tools for using Julia with GPUs are worth a look as well. In this blogpost, I will
+list some of those packages but focus on the latest addition: native GPU programming with
+CUDAnative.jl
+
+
+
+## Available tools
+
+The Julia package ecosystem contains quite a few GPU-related packages, targeting different
+levels of abstraction. At the highest abstraction level, domain-specific packages like
+[MXNet.jl](https://github.com/dmlc/MXNet.jl) or
+[TensorFlow.jl](https://github.com/malmaud/TensorFlow.jl) can transparently use any
+available GPUs. More generic development is possible with
+[ArrayFire.jl](https://github.com/JuliaComputing/ArrayFire.jl), and if you need a
+specialized implementation you can use wrappers for vendor-specific library packages
+(cuBLAS, cuFFT, cuSPARSE, ...) available at the [JuliaGPU GitHub
+organization](https://github.com/JuliaGPU/). Each of those packages is just a single
+`Pkg.add(...)` away!
+
+What's missing from this list is the lowest abstraction level, where you can create kernels
+and manage execution like you would in CUDA. This is where CUDAnative.jl comes in.
+
+
+
+## Native GPU programming with CUDAnative.jl
+
+[CUDAnative.jl](https://github.com/JuliaGPU/CUDAnative.jl) is a brand-new package adding
+native GPU programming capabilities to the Julia programming language. Together with a
+package like [CUDAdrv.jl](https://github.com/JuliaGPU/CUDAdrv.jl) or
+[CUDArt.jl](https://github.com/JuliaGPU/CUDArt.jl) for interfacing with the CUDA libraries,
+it is now possible to do CUDA development in Julia without the need for an external language
+or compiler. As an introductory example, the following listing shows how to compute the sum
+of two vectors:
+
+```julia
+using CUDAdrv, CUDAnative
+
+function kernel_vadd(a, b, c)
+    i = (blockIdx().x-1) * blockDim().x + threadIdx().x
+    c[i] = a[i] + b[i]
+
+    return nothing
+end
+
+dev = CuDevice(0)
+ctx = CuContext(dev)
+
+# generate some data
+len = 512
+a = rand(Int, len)
+b = rand(Int, len)
+
+# allocate & upload on the GPU
+d_a = CuArray(a)
+d_b = CuArray(b)
+d_c = similar(d_a)
+
+# execute and fetch results
+@cuda (1,len) kernel_vadd(d_a, d_b, d_c)
+c = Array(d_c)
+
+destroy(ctx)
+```
+
+The real workhorse of this example is the `@cuda` macro, which generates specialized code
+for compiling the kernel function to GPU assembly, uploading it to the driver, and preparing
+the execution environment. Together with Julia's JIT compiler, this results in a very
+efficient kernel launch sequence, avoiding the runtime overhead typically associated with
+dynamic languages.
+
+<!-- This seems a bit short. Maybe elaborate more? -->
```

Comment:
Also worth putting some kind of link in so that folks who don't already know about `blockIdx` etc can go read about it. Is it worth commenting on why `blockIdx` is a function rather than a constant?